palace/models/curlcurloperator.cpp

```diff
@@ -106,79 +113,78 @@ void CurlCurlOperator::CheckBoundaryProperties()
+    MaterialPropertyCoefficient<MatType> muinv_func(mat_op);
+    auto k = std::make_unique<mfem::SymmetricBilinearForm>(&nd_fespace_l);
+    k->AddDomainIntegrator(new mfem::CurlCurlIntegrator(muinv_func));
+    k->SetAssemblyLevel(assembly_level);
+    k->Assemble(skip_zeros);
+    k->Finalize(skip_zeros);
+    if (print_hdr)
     {
-      std::string str = "";
-      if (pc_gmg)
+      Mpi::Print(" Level {:d}: {:d} unknowns", l, nd_fespace_l.GlobalTrueVSize());
+      if (assembly_level == mfem::AssemblyLevel::LEGACY)
+      {
+        HYPRE_BigInt nnz = k->SpMat().NumNonZeroElems();
+        Mpi::GlobalSum(1, &nnz, nd_fespace_l.GetComm());
+        Mpi::Print(", {:d} NNZ\n", nnz);
+      }
+      else
       {
-        str =
-            fmt::format(" (Level {:d}, {:d} unknowns)", l, nd_fespace_l.GlobalTrueVSize());
+        Mpi::Print("\n");
       }
-      Mpi::Print(" K{}: NNZ = {:d}, norm = {:e}\n", str, hK->NNZ(),
-                 hypre_ParCSRMatrixFnorm(*hK));
     }
-    K.emplace_back(hK);
+    auto K_l = std::make_unique<ParOperator>(std::move(k), nd_fespace_l);
+    K_l->SetEssentialTrueDofs(dbc_tdof_lists[l], Operator::DiagonalPolicy::DIAG_ONE);
+    K->AddOperator(std::move(K_l));
   }
+  print_hdr = false;
+  return K;
 }
 
-std::unique_ptr<mfem::Operator> CurlCurlOperator::GetCurlMatrix()
+std::unique_ptr<Operator> CurlCurlOperator::GetCurlMatrix()
 {
-  mfem::ParDiscreteLinearOperator curl(&nd_fespaces.GetFinestFESpace(), &rt_fespace);
-  curl.AddDomainInterpolator(new mfem::CurlInterpolator);
-  // curl.SetAssemblyLevel(mfem::AssemblyLevel::FULL);
-  curl.Assemble();
-  curl.Finalize();
-  return std::unique_ptr<mfem::HypreParMatrix>(curl.ParallelAssemble());
+  auto curl = std::make_unique<mfem::DiscreteLinearOperator>(&GetNDSpace(), &GetRTSpace());
+  curl->AddDomainInterpolator(new mfem::CurlInterpolator);
+  curl->SetAssemblyLevel(assembly_level);
+  curl->Assemble();
+  curl->Finalize();
+  return std::make_unique<ParOperator>(std::move(curl), GetNDSpace(), GetRTSpace(), true);
 }
 
-void CurlCurlOperator::GetExcitationVector(int idx, mfem::Vector &RHS)
+void CurlCurlOperator::GetExcitationVector(int idx, Vector &RHS)
 {
   // Assemble the surface current excitation +J. The SurfaceCurrentOperator assembles -J
   // (meant for time or frequency domain Maxwell discretization, so we multiply by -1 to
   // retrieve +J).
-  SumVectorCoefficient fb(nd_fespaces.GetFinestFESpace().GetParMesh()->SpaceDimension());
+  SumVectorCoefficient fb(GetNDSpace().GetParMesh()->SpaceDimension());
   surf_j_op.AddExcitationBdrCoefficients(idx, fb);
-  RHS.SetSize(nd_fespaces.GetFinestFESpace().GetTrueVSize());
+  RHS.SetSize(GetNDSpace().GetTrueVSize());
   RHS = 0.0;
   if (fb.empty())
   {
     return;
   }
-  mfem::ParLinearForm rhs(&nd_fespaces.GetFinestFESpace());
+  mfem::LinearForm rhs(&GetNDSpace());
   rhs.AddBoundaryIntegrator(new VectorFEBoundaryLFIntegrator(fb));
-  rhs.UseFastAssembly(true);
+  rhs.UseFastAssembly(false);
```

Comment:
The fast assembly should not have been active before because it isn't supported for the linear forms we are using or for mixed/tet meshes. So I am just turning it off explicitly to be safe.